astropy/tests/runner.py

```diff
@@ -1,141 +1,135 @@
-            When `True`, skips running the doctests in the .rst files.
-
-        repeat : int, optional
-            If set, specifies how many times each test should be run. This is
-            useful for diagnosing sporadic failures.
-
-        See Also
-        --------
-        pytest.main : py.test function wrapped by `run_tests`.
+        Before constructing the class parse all the methods that have been
+        decorated with ``keyword``.
+
+        The objective of this method is to construct a default set of keyword
+        arguments to the ``run_tests`` method. It does this by inspecting the
+        methods of the class for functions with the name ``keyword`` which is
+        the name of the decorator wrapping function. Once it has created this
+        dictionary, it also formats the docstring of ``run_tests`` to be
+        comprised of the docstrings for the ``keyword`` methods.
+
+        To add a keyword argument to the ``run_tests`` method, define a new
+        method decorated with ``@keyword`` and with the ``self, name, kwargs``
+        signature.
         """
+        # Get all 'function' members as the wrapped methods are functions
+        if six.PY2:
+            functions = inspect.getmembers(cls, predicate=inspect.ismethod)
+        else:
+            functions = inspect.getmembers(cls, predicate=inspect.isfunction)
+
+        # Filter out anything that's not got the name 'keyword'
+        keywords = filter(lambda func: func[1].__name__ == 'keyword', functions)
+        # Sort all keywords based on the priority flag.
+        sorted_keywords = sorted(keywords, key=lambda x: x[1]._priority, reverse=True)
+
+        cls.keywords = OrderedDict()
+        doc_keywords = ""
+        for name, func in sorted_keywords:
+            cls.keywords[name] = func._default_value
+            if func.__doc__:
+                doc_keywords += func.__doc__
+        if six.PY2:
+            cls.run_tests.__func__.__doc__ = cls.run_tests.__doc__.format(keywords=doc_keywords)
+        else:
+            cls.run_tests.__doc__ = cls.run_tests.__doc__.format(keywords=doc_keywords)
 
-        # Don't import pytest until it's actually needed to run the tests
-        from .helper import pytest
+        return super(TestRunner, cls).__new__(cls)
 
+    @keyword(None, priority=100)
+    def coverage(self, coverage, kwargs):
         if coverage:
             warnings.warn(
                 "The coverage option is ignored on run_tests, since it "
                 "can not be made to work in that context.  Use "
                 "'python setup.py test --coverage' instead.",
                 AstropyWarning)
 
-        all_args = []
+        return []
 
+    @keyword(None, priority=1)
+    def package(self, package, kwargs):
+        """
+        package : str, optional
+            The name of a specific package to test, e.g. 'io.fits' or 'utils'.
+            If nothing is specified all default Astropy tests are run.
+
```

Comment:
What about not putting extraneous spaces here and then in ``__new__`` adding spaces if needed between parameters?